New TransportBroadcastUnpromotableAction action

kingherc opened the pull request:
Introduces:

* New action that can be used to broadcast to unpromotable shards of a given primary shard.
* New hook in ReplicationOperation for custom logic when the primary operation completes.
* Refresh action now uses the new hook to broadcast the unpromotable refresh action to all unpromotable shards.

Fixes ES-5454

kingherc committed:
New TransportBroadcastUnpromotableAction action

Introduces:

* New action that can be used to broadcast to unpromotable
  shards of a given primary shard.
* New hook in ReplicationOperation for custom logic when the
  primary operation completes.
* Refresh action now uses the new hook to broadcast the
  unpromotable refresh action to all unpromotable shards.

Fixes ES-5454

kingherc committed:
Update docs/changelog/93600.yaml

elasticsearchmachine commented:
Hi @kingherc, I've created a changelog YAML for you.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java:
```diff
@@ -0,0 +1,112 @@
+    protected void doExecute(Task task, Request request, ActionListener<ActionResponse.Empty> listener) {
+        try (var listeners = new RefCountingListener(listener.map(v -> ActionResponse.Empty.INSTANCE))) {
+            ActionListener.completeWith(listeners.acquire(), () -> {
+                final ClusterState clusterState = clusterService.state();
```
I think we should carry the shard routing table in the request rather than getting it again from here.

kingherc replied:
Hi @DaveCTurner , I have not got this point yet, why? In the cases we have so far, both the refresh and new commit notification call this new action locally, so I do not see why the shard routing table would be different from the cluster state's one. Plus, while the Refresh one calls it internally from within another action, the new commit notification does not need to do that anymore (we save one round-trip to the local node of the primary) and can call this action directly.

DaveCTurner replied:
`clusterService#state()` is mutable, it returns the last-applied state, so it may have changed since the last time we read it during the coordination of the refresh action. Mostly this will be fine, but sometimes it's going to yield a totally different routing table (e.g. the index may even be deleted at this point). I'd prefer the caller to have control over which routing table to use for the routing of this request. The new-commit notification can just get it from the cluster service for sure.

kingherc replied:
Hi @DaveCTurner I see what you mean. But in `TransportShardRefreshAction.UnpromotableReplicasRefreshProxy#onPrimaryOperationComplete()` we'd need to get hold of a shard routing table to pass in the Request. Wouldn't we still get that from the cluster service? Or does Refresh somehow holds the primary's shard routing table?

DaveCTurner replied:
it comes from the `ReplicationGroup`:

```diff
diff --git a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java
index b5d805f270c..9ab8716ce4a 100644
--- a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java
@@ -17,6 +17,7 @@ import org.elasticsearch.action.support.replication.ReplicationResponse;
 import org.elasticsearch.action.support.replication.TransportReplicationAction;
 import org.elasticsearch.client.internal.node.NodeClient;
 import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.common.io.stream.StreamInput;
@@ -108,11 +109,16 @@ public class TransportShardRefreshAction extends TransportReplicationAction<
     protected class UnpromotableReplicasRefreshProxy extends ReplicasProxy {
 
         @Override
-        public void onPrimaryOperationComplete(BasicReplicationRequest replicaRequest, ActionListener<Void> listener) {
+        public void onPrimaryOperationComplete(
+            BasicReplicationRequest replicaRequest,
+            IndexShardRoutingTable indexShardRoutingTable,
+            ActionListener<Void> listener
+        ) {
             assert primaryRefreshResult.refreshed() : "primary has not refreshed";
             ShardId primary = replicaRequest.shardId();
             UnpromotableShardRefreshRequest unpromotableReplicaRequest = new UnpromotableShardRefreshRequest(
                 primary,
+                indexShardRoutingTable,
                 primaryRefreshResult.generation()
             );
             client.executeLocally(
diff --git a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java
index bd302a3dc6e..31416d77178 100644
--- a/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java
@@ -10,6 +10,7 @@ package org.elasticsearch.action.admin.indices.refresh;
 
 import org.elasticsearch.action.ActionRequestValidationException;
 import org.elasticsearch.action.support.broadcast.unpromotable.BroadcastUnpromotableRequest;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.index.engine.Engine;
@@ -23,8 +24,12 @@ public class UnpromotableShardRefreshRequest extends BroadcastUnpromotableReques
 
     private final long segmentGeneration;
 
-    public UnpromotableShardRefreshRequest(final ShardId primaryShardId, long segmentGeneration) {
-        super(primaryShardId);
+    public UnpromotableShardRefreshRequest(
+        final ShardId primaryShardId,
+        IndexShardRoutingTable indexShardRoutingTable,
+        long segmentGeneration
+    ) {
+        super(primaryShardId, indexShardRoutingTable);
         this.segmentGeneration = segmentGeneration;
     }
 
diff --git a/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java b/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java
index ceef84c2026..8117226529e 100644
--- a/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java
+++ b/server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java
@@ -10,6 +10,7 @@ package org.elasticsearch.action.support.broadcast.unpromotable;
 
 import org.elasticsearch.action.ActionRequest;
 import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.index.shard.ShardId;
@@ -24,14 +25,16 @@ import static org.elasticsearch.action.ValidateActions.addValidationError;
 public class BroadcastUnpromotableRequest extends ActionRequest {
 
     protected final ShardId primaryShardId;
+    protected final IndexShardRoutingTable indexShardRoutingTable;
 
     public BroadcastUnpromotableRequest(StreamInput in) throws IOException {
-        super(in);
-        primaryShardId = new ShardId(in);
+        assert false;
+        throw new UnsupportedOperationException("not supported");
     }
 
-    public BroadcastUnpromotableRequest(ShardId primaryShardId) {
+    public BroadcastUnpromotableRequest(ShardId primaryShardId, IndexShardRoutingTable indexShardRoutingTable) {
         this.primaryShardId = primaryShardId;
+        this.indexShardRoutingTable = indexShardRoutingTable;
     }
 
     public ShardId primaryShardId() {
diff --git a/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java b/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
index 98e1f02c324..f814cb915e1 100644
--- a/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
+++ b/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
@@ -133,9 +133,12 @@ public class ReplicationOperation<
                 logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
             }
 
+            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
+
             pendingActions.incrementAndGet();
             replicasProxy.onPrimaryOperationComplete(
                 replicaRequest,
+                replicationGroup.getRoutingTable(),
                 ActionListener.wrap(ignored -> decPendingAndFinishIfNeeded(), e -> { finishAsFailed(e); })
             );
 
@@ -152,7 +155,6 @@ public class ReplicationOperation<
             // on.
             final long maxSeqNoOfUpdatesOrDeletes = primary.maxSeqNoOfUpdatesOrDeletes();
             assert maxSeqNoOfUpdatesOrDeletes != SequenceNumbers.UNASSIGNED_SEQ_NO : "seqno_of_updates still uninitialized";
-            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
             final PendingReplicationActions pendingReplicationActions = primary.getPendingReplicationActions();
             markUnavailableShardsAsStale(replicaRequest, replicationGroup);
             performOnReplicas(replicaRequest, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, replicationGroup, pendingReplicationActions);
@@ -615,7 +617,11 @@ public class ReplicationOperation<
          * @param replicaRequest             the operation that will be performed on replicas
          * @param listener                   callback for handling the response or failure
          */
-        default void onPrimaryOperationComplete(RequestT replicaRequest, ActionListener<Void> listener) {
+        default void onPrimaryOperationComplete(
+            RequestT replicaRequest,
+            IndexShardRoutingTable indexShardRoutingTable,
+            ActionListener<Void> listener
+        ) {
             listener.onResponse(null);
         }
     }
diff --git a/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java b/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java
index 553b212d83d..6391ac591f5 100644
--- a/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java
@@ -14,6 +14,7 @@ import org.elasticsearch.action.support.ActionFilters;
 import org.elasticsearch.action.support.ActionTestUtils;
 import org.elasticsearch.action.support.PlainActionFuture;
 import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
 import org.elasticsearch.cluster.routing.ShardRouting;
 import org.elasticsearch.cluster.routing.ShardRoutingState;
 import org.elasticsearch.cluster.service.ClusterService;
@@ -127,8 +128,8 @@ public class TransportBroadcastUnpromotableActionTests extends ESTestCase {
             super(in);
         }
 
-        TestBroadcastUnpromotableRequest(ShardId primaryShardId) {
-            super(primaryShardId);
+        TestBroadcastUnpromotableRequest(ShardId primaryShardId, IndexShardRoutingTable indexShardRoutingTable) {
+            super(primaryShardId, indexShardRoutingTable);
         }
 
     }
@@ -178,7 +179,10 @@ public class TransportBroadcastUnpromotableActionTests extends ESTestCase {
             ActionTestUtils.execute(
                 broadcastUnpromotableAction,
                 null,
-                new TestBroadcastUnpromotableRequest(primaryShardId.shardId()),
+                new TestBroadcastUnpromotableRequest(
+                    primaryShardId.shardId(),
+                    state.routingTable().shardRoutingTable(primaryShardId.shardId())
+                ),
                 response
             );
         });
@@ -263,7 +267,12 @@ public class TransportBroadcastUnpromotableActionTests extends ESTestCase {
         ShardId wrongShardId = new ShardId(shardId.getIndex(), shardId.getId() + randomIntBetween(10, 100));
         PlainActionFuture<ActionResponse.Empty> response = PlainActionFuture.newFuture();
         logger.debug("--> executing for primary shard id: {}", wrongShardId);
-        ActionTestUtils.execute(broadcastUnpromotableAction, null, new TestBroadcastUnpromotableRequest(wrongShardId), response);
+        ActionTestUtils.execute(
+            broadcastUnpromotableAction,
+            null,
+            new TestBroadcastUnpromotableRequest(wrongShardId, state.routingTable().shardRoutingTable(wrongShardId)),
+            response
+        );
 
         ExecutionException e = expectThrows(ExecutionException.class, () -> response.get());
         assertThat(e, hasToString(containsString("no such shard")));

```

DaveCTurner replied:
We'd still be missing the nodes here - I think it would be reasonable to capture them in the `ReplicationGroup` too, but I would also be ok with grabbing them from the cluster service, with suitable protections for the case where a node doesn't exist.

kingherc replied:
Thanks for the explanation @DaveCTurner . I did not realize ReplicationGroup had the routing table. I guess we'd need to also make IndexShardRoutingTable Writeable. I saw above you did not make it, but I think it needs to be, right? Or do you mean we don't need to transmit it, and it's only active on the local request and then not passed through the wire requests (making it null)?

DaveCTurner replied:
I think we can assert that we never serialise this, it's only ever used locally. It is already kinda writeable via `org.elasticsearch.cluster.routing.IndexShardRoutingTable.Builder#writeTo` but this isn't important. I'm talking with the core/infra folks about dropping the need for these local-only requests to implement `TransportRequest`.

kingherc replied:
Hi @DaveCTurner , it's only used locally but the Request is broadcasted to the unpromotable shards. For example in the case of UnpromotableShardRefreshRequest it has the segmentGeneration. So the whole Request (inheriting from BroadcastUnpromotableRequest) needs to be able to be  serialized. We could drop though the IndexShardRoutingTable from the serialization, thus it becomes null across the wire. Is that fine?

DaveCTurner replied:
Oh ok I see. Hmm. We should really have a different request class for the ones that do go over the wire. I can live with the routing table just being dropped tho.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java:
```diff
@@ -10,38 +10,47 @@
-    public static final String NAME = RefreshAction.NAME + "[u]";
+public class TransportUnpromotableShardRefreshAction extends TransportBroadcastUnpromotableAction<UnpromotableShardRefreshRequest> {
+
+    public static final String NAME = "indices:admin/refresh/unpromotable";
```
nit: maybe let's mention `RefreshAction.NAME` here still just to link them together a bit more?

```suggestion
    public static final String NAME = RefreshAction.NAME + "/unpromotable";
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java:
```diff
@@ -0,0 +1,112 @@
+
+        @Override
+        public void messageReceived(Request request, TransportChannel channel, Task task) throws Exception {
+            final ActionListener<ActionResponse.Empty> listener = ActionListener.wrap(channel::sendResponse, e -> {
```
I'd prefer to use a `ChannelActionListener` here

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -132,6 +132,13 @@ private void handlePrimaryResult(final PrimaryResultT primaryResult) {
+            pendingActions.incrementAndGet();
+            replicasProxy.onPrimaryOperationComplete(
+                replicaRequest,
+                ActionListener.wrap(ignored -> decPendingAndFinishIfNeeded(), e -> { finishAsFailed(e); })
```
nit: I think we can write this? (haven't tried tho)

```suggestion
                ActionListener.wrap(ignored -> decPendingAndFinishIfNeeded(), this::finishAsFailed)
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableActionTests.java:
```diff
@@ -0,0 +1,272 @@
+
+        ShardId shardId = state.routingTable().activePrimaryShardsGrouped(new String[] { index }, true).get(0).shardId();
+        ShardId wrongShardId = new ShardId(shardId.getIndex(), shardId.getId() + randomIntBetween(10, 100));
+        PlainActionFuture<ActionResponse.Empty> response = PlainActionFuture.newFuture();
+        logger.debug("--> executing for primary shard id: {}", wrongShardId);
+        ActionTestUtils.execute(broadcastUnpromotableAction, null, new TestBroadcastUnpromotableRequest(wrongShardId), response);
```
nit: maybe use the static `PlainActionFuture#get` for things like this (ideally one with a timeout, there's no need to wait forever here)

kingherc replied:
It's done in the line below with `response.get()`. Feel free to tell me if this is not what you meant.

kingherc replied:
Feel free to unresolve the conversation.

DaveCTurner replied:
I mean the _static_ `get` method:

```
        logger.debug("--> executing for primary shard id: {}", wrongShardId);
        assertThat(
            expectThrows(
                ShardNotFoundException.class,
                () -> PlainActionFuture.<ActionResponse.Empty, Exception>get(
                    f -> ActionTestUtils.execute(broadcastUnpromotableAction, null, new TestBroadcastUnpromotableRequest(wrongShardId), f),
                    10,
                    TimeUnit.SECONDS
                )
            ).toString(),
            allOf(containsString("no such shard"), containsString(wrongShardId.toString()))
        );
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -709,6 +710,7 @@ public <Request extends ActionRequest, Response extends ActionResponse> void reg
         actions.register(TransportShardFlushAction.TYPE, TransportShardFlushAction.class);
         actions.register(TransportShardRefreshAction.TYPE, TransportShardRefreshAction.class);
         actions.register(TransportPrevalidateShardPathAction.TYPE, TransportPrevalidateShardPathAction.class);
+        actions.register(TransportUnpromotableShardRefreshAction.TYPE, TransportUnpromotableShardRefreshAction.class);
```
Hmm I don't think we should register this here because I see no need for a client to call it directly (unless I'm missing something). Instead, we can register it in the constructor of `TransportShardRefreshAction`.

DaveCTurner reviewed (commented):
I like it 👍 I did an initial pass and left a few (mostly superficial) comments.

kingherc committed:
Fix style

kingherc committed:
Fix PR comments

kingherc committed:
Remove ActionType

kingherc committed:
Fix style

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java:
```diff
@@ -11,37 +11,44 @@
-    public static final String NAME = RefreshAction.NAME + "[u]";
+public class TransportUnpromotableShardRefreshAction extends TransportBroadcastUnpromotableAction<UnpromotableShardRefreshRequest> {
+
+    public static final String NAME = RefreshAction.NAME + "/unpromotable";
```
I have a doubt so I prefer to comment: does this new action require specific privileges? IIRC it is ok as long as the original action has been authorized but I'd be happy if someone can confirm because I see `indices:admin/refresh[s]` and `indices:admin/refresh*` privileges in the codebase.

kingherc replied:
Hi @tlrx ! By privileges, what do you mean? Do you mean the `x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java` file? We did not need to put it there, because of [this comment](https://github.com/elastic/elasticsearch/pull/93600#discussion_r1100488177) whereby now this new action is internal and not exposed to clients I believe.

kingherc replied:
I think I can rename the action to what we have currently in `main`, i.e., `indices:admin/refresh[s][u]`. or maybe better `indices:admin/refresh[u]`. This followed a similar approach to how the replication messages were done, that append a `[r]` to the action name. Let me see if that resolves the CI issues.

tlrx replied:
The `RBACEngine#authorizeIndexAction` engine is doing a bunch of verifications depending on the action name (that starts with `indices:`) and if the UnpromotableShardRefreshRequest is an instance or not of `IndicesRequest`. 

This is not the case so the only known actions that contains `indices:` and are not `IndicesRequest` are scroll and async search it seems.

That being said we must be compliant with the RBACEngine or adjust its checks, but I don't think we need new privileges.

kingherc replied:
I guess I got confused because currently in `main` we do call the same action (see [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java#L110-L120)) without an RBAC error. So I am not sure why this issue was created now.

I just pushed a fix for RBACEngine rules. Feel free to review as I am not aware of that code piece.

tlrx replied:
It used `transportService.sendChildRequest()` which is `transportService.sendRequest()` in this PR

kingherc replied:
I tried that also but it did not work locally. Are you sure that was the difference? Because i can try that again if that's preferred?

kingherc replied:
I will resolve this conversation, but feel free to unresolve it during the main reviewing or comment in the part with the RBACEngine changes!

kingherc committed:
Rename action

kingherc committed:
Pass IndexShardRoutingTable to action

kingherc committed:
Fix RBACEngine checks

kingherc committed:
Merge remote-tracking branch 'main' into feature/ES-5454-action-broadcast-to-unpromotables

kingherc committed:
Fix checkstyle

kingherc committed:
Fix leftover TODOs

kingherc committed:
Fix test

kingherc commented:
@elasticmachine run elasticsearch-ci/part-2

kingherc commented:
This is now ready for review @DaveCTurner , @tlrx , @idegtiarenko . Thanks for any feedback!

kingherc committed:
Rename primaryShardId to shardId

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/BroadcastUnpromotableRequest.java:
```diff
@@ -23,15 +26,24 @@
+     * the unpromotable replicas. The routing table is not serialized over the wire, and will be null on the other end of the wire.
+     * For this reason, the field is package-private.
+     */
+    final @Nullable IndexShardRoutingTable indexShardRoutingTable;
```
Should it be private?

kingherc replied:
Hi @idegtiarenko , no, because it's used `TransportBroadcastUnpromotableAction#doExecute` and that is why I wanted it to be package-private (as mentioned in the comment). Feel free to review the comment.

DaveCTurner replied:
There's not much more to an `IndexShardRoutingTable` than just the `ShardRouting` things it contains. I'd rather keep the whole ISRT here - if we don't have the `ShardRouting` objects then we can't filter out the promotable shards within the action, we have to ask every caller to do that, and the whole point of this PR is to centralise that logic.

kingherc replied:
I also believe we may be more flexible with the whole IndexShardRoutingTable in the future if ever needed.

kingherc replied:
@tlrx You mean that the cluster state might already be updated on the local primary node during the replication action, and we can now detect it here. But what would happen if it's not of a higher version? Would it throw or proceed somehow with the latest cluster state? In this case, it proceeds with the index shard routing table of the older version. If it would proceed with the most recent cluster state, then we'd end up not using the ISRT that @DaveCTurner wanted. If it would throw, then that's a new behavior that I can implement, but I wonder if we want that? It would mean that the promotable replicas might proceed (with the older replication group) while the unpromotables would fail because we're not on the latest cluster state. I find this discrepancy a bit weird.

kingherc replied:
From @tlrx :

> I've been puzzled by the fact that we want to reuse the unpromotable shards from the same replication group as the original replication action (and hence we must locally pass the ISRT around)... but a responsability of replication action is to mark shards as stale, ie removed them from the in sync set, and unpromotable shards do not belong to the in-sync set and as such why does this matter to use the same ISRT => hence my question if a "fresh" enough ISRT could be enough. But I got the point of to keep it close of the TransportReplicationAction, I see how it can help for future action. I'm not going to slow you down more, thanks both for answering.

@DaveCTurner I also saw your answer.

I think from your comments you both do not require anything additional in the PR at the moment in this aspect. But feel free to tell me if I misunderstood and you prefer the way of somehow checking the cluster state version.

Thanks!

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java:
```diff
@@ -72,7 +70,7 @@ public TransportShardRefreshAction(
             ThreadPool.Names.REFRESH
         );
-        new TransportUnpromotableShardRefreshAction(transportService, actionFilters, indicesService);
+        new TransportUnpromotableShardRefreshAction(clusterService, transportService, actionFilters, indicesService);
```
```suggestion
        // registers the unpromotable version of shard refresh action
        new TransportUnpromotableShardRefreshAction(clusterService, transportService, actionFilters, indicesService);
```

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java:
```diff
@@ -86,43 +84,11 @@ protected void shardOperationOnPrimary(
-        try (var listeners = new RefCountingListener(listener.map(v -> new PrimaryResult<>(shardRequest, new ReplicationResponse())))) {
-            var refreshResult = primary.refresh(SOURCE_API);
+        ActionListener.completeWith(listener, () -> {
+            primaryRefreshResult = primary.refresh(SOURCE_API);
```
Transport actions are instantiated as singletons so storing a result specific to a shard in a class attribute can cause bugs. That would be nice if we can use the `PrimaryResult` to hold the primary shard result for the time of the execution

kingherc replied:
Ah good catch! I just realize this. I did not want originally to be touching the replica request (to avoid any bwc issues), but I see now it is necessary. However, I can do something similar we did with the IndexShardRoutingTable parameter in the unpromotable request -- make the primary refresh result a package-private parameter to add in the replica request, and not send/serialize it over the wire. I did that in the latest commit, feel free to review that.

tlrx reviewed (commented):
I had a first look and I have comments

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java:
```diff
@@ -0,0 +1,102 @@
+        super(actionName, transportService, actionFilters, requestReader);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.transportUnpromotableAction = actionName + "[u]";
```
That's the shard-level action name, yes.

tlrx replied:
Thanks :+1:

kingherc commented on server/src/main/java/org/elasticsearch/action/admin/indices/refresh/UnpromotableShardRefreshRequest.java:
```diff
@@ -8,52 +8,52 @@
     @Override
     public String toString() {
-        return "UnpromotableShardRefreshRequest{" + "shardId=" + shardId + ", segmentGeneration=" + segmentGeneration + '}';
+        return "UnpromotableShardRefreshRequest{" + "shardId=" + shardId() + ", segmentGeneration=" + segmentGeneration + '}';
```
Hi @tlrx . In general, I followed the approach of other requests I saw in the code where their description is usually the class name and the fields of the request. What would you like to see? Note that the base class's getDescription() of this class uses the toString() method.

kingherc committed:
Make refresh result part of the replica request

Plus some other PR comments resolved

kingherc committed:
Do not fork again

kingherc committed:
Merge remote-tracking branch 'main' into feature/ES-5454-action-broadcast-to-unpromotables

kingherc reviewed (commented):
Thank you all for the feedback! I handled it, please feel free to take another look. Thanks.

kingherc committed:
Fix merge conflict

kingherc commented:
part-1 error is https://github.com/elastic/elasticsearch/issues/93745 which I just opened. Retrying test.

kingherc commented:
@elasticmachine run elasticsearch-ci/part-1

kingherc committed:
Merge remote-tracking branch 'main' into feature/ES-5454-action-broadcast-to-unpromotables

kingherc commented:
Opened https://github.com/elastic/elasticsearch/issues/93748 for another CI error. Retrying

kingherc commented:
@elasticmachine run elasticsearch-ci/part-3

kingherc committed:
Fail if primary customization logic fails

Fixes ES-5212

kingherc commented:
Hi reviewers, added the last commit that resolves also [ES-5212](https://elasticco.atlassian.net/browse/ES-5212?focusedCommentId=39355) based on the linked slack thread conversation there. It means that if there is a failure during the refresh of the unpromotable shards, this results in failure(s) of the refresh request. Added a relevant test.

A reminder as well if you can to review as soon as possible :)

kingherc committed:
Register new exception

kingherc committed:
Add new exception to test

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -132,6 +133,18 @@ private void handlePrimaryResult(final PrimaryResultT primaryResult) {
+                replicationGroup.getRoutingTable(),
+                ActionListener.wrap(
+                    ignored -> decPendingAndFinishIfNeeded(),
+                    exception -> this.finishAsFailed(new OnPrimaryOperationCompleteException(primary.routingEntry().shardId(), exception))
```
Hmm I don't think we should `finishAsFailed` here - we should treat failures here similarly to replica failures, which involves adding them to `shardReplicaFailures` but allowing the overall operation to succeed.

kingherc replied:
Hi @DaveCTurner , I do not think we can do that, as there is no way we can know which shard/node failed. `onPrimaryOperationComplete()` is a general customization function that does not operate on a single shard/replica, but the user is able to do whatever they would like. Thus, I believe the `finishAsFailed` is better. 

> but allowing the overall operation to succeed.

I thought the overall strategy we wanted was to fail the refresh if the unpromotable refresh fails. So this is another reason I chose `finishAsFailed`.

If we would like it to still make progress, then we could:
* Either revert my latest commit, which would bubble up the previous exception and the primary would retry with success (after removing the disconnected node), but the Refresh would then succeed.
* We could make more infrastructure to somehow make `onPrimaryOperationComplete()` on a per-shard level. But then we go back to my idea of integrating the unpromotable replication inside ReplicationOperation rather than leaving it to the implementor to do that outside of it.

Please advise :)

DaveCTurner replied:
I believe we can create a `org.elasticsearch.action.support.replication.ReplicationResponse.ShardInfo.Failure` representing the failure - the `nodeId` field is (or at least claims to be) optional.

If not, I'd rather just log and drop the exception for now and come back to the question of failure-handling in a follow-up.

kingherc replied:
You're right, thanks for the tip! I managed to make it behave like a replica shard failure. Feel free to see the relevant commit when it's pushed and comment :)

DaveCTurner commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -343,7 +344,7 @@ public void authorizeIndexAction(
                     listener.onResponse(IndexAuthorizationResult.ALLOW_NO_INDICES);
                 }
-            } else if (action.equals(ClosePointInTimeAction.NAME)) {
+            } else if (action.equals(ClosePointInTimeAction.NAME) || action.equals(TransportUnpromotableShardRefreshAction.NAME)) {
```
I don't think we should add another security bypass like this. The user should be authorised to refresh the indices to which they're writing, and if they aren't then we should respect that.

kingherc replied:
This is resolved with the new commit I will push. There's no such exception anymore.

DaveCTurner replied:
Still seems to be the case - let's keep this unresolved until the fix is pushed.

kingherc replied:
@DaveCTurner done!

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -43,21 +43,22 @@
-    final List<ShardRouting> replicas;
-    final List<ShardRouting> activeShards;
-    final List<ShardRouting> assignedShards;
+    private final ShardShuffler shuffler;
```
I'd prefer that we do this in a follow-up to reduce noise in this PR and make it a bit easier to review.

kingherc replied:
Ehm, this was a review comment from @idegtiarenko I'm trying to handle. I can revert this if it's OK for you @idegtiarenko ?

DaveCTurner replied:
I see, yes. It's a valid comment, but best to keep these things separate where possible

kingherc replied:
@DaveCTurner I reverted these changes. @idegtiarenko feel free to tell us whether that's not OK for you (I can re-revert it).

DaveCTurner reviewed (commented):
I left a few more comments

kingherc commented on server/src/main/java/org/elasticsearch/transport/TransportService.java:
```diff
@@ -725,7 +725,6 @@ public final <T extends TransportResponse> void sendRequest(
             return;
         } catch (Exception exception) {
             // shouldn't happen in practice, but handle it anyway to be sure
-            assert false : exception;
```
Hi @DaveCTurner , I removed this in order to be able to do the `TransportBroadcastUnpromotableActionTests#testInvalidNodes` test. I think we said to test if there were invalid nodes in the index shard routing table. But the test then trips here. If this branch cannot happen in practice, then I guess I will remove the test.

DaveCTurner replied:
ah I think I see. Unknown nodes should yield something like a `NodeNotConnectedException` which would be handled by the other catch block.

kingherc replied:
@DaveCTurner , true about the other catch block, but in my case the testInvalidNodes results in a `java.lang.AssertionError: java.lang.NullPointerException: discovery node must not be null`. Because `TransportBroadcastUnpromotableAction#doExecute` does `final DiscoveryNode node = clusterState.nodes().get(shardRouting.currentNodeId());` (with a wrong node id due to the test) and this results in null which is then passed to the sendRequest method whose `isLocalNode()` requires a non-null node id.

DaveCTurner replied:
Yes, I think we should map this to a `NodeNotConnectedException` somehow. This is definitely possible because `clusterState.nodes()` comes from a different cluster state from the ISRT so may indeed be missing nodes.

kingherc replied:
OK I see, I did that, thanks!

kingherc committed:
Fail unpromotable like replica shard failures

Fixes ES-5212

kingherc committed:
Null node throws NodeNotConnectedException

kingherc commented:
Hi @tlrx , @idegtiarenko , @DaveCTurner , I handled the feedback so far. Feel free to review again, thanks!

kingherc committed:
Revert changing private fields in routing class

kingherc commented:
I am not sure @tlrx about your comment on ISRT . @davecturner suggested it should be the responsibility of the user to pass the index shard routing table rather than the action inferring it from the cluster state. You suggest to check the cluster state number, but I am not sure how that would work? This is already executed locally on the primary node, so it would mean to check that the primary node's cluster state is already out of sync? how would it know if the number is higher or not? I feel we have the same problem again. Or I am mistaken in my thinking.

kingherc commented:
@tlrx 

> as long as it works on a cluster state version higher than the one that executed the original replication action?

Ah I now see. You mean that the cluster state might already be updated on the local primary node during the replication action, and we can now detect it here. But what would happen if it's not of a higher version? Would it throw or proceed somehow with the latest cluster state? In this case, it proceeds with the index shard routing table of the older version. If it would proceed with the most recent cluster state, then we'd end up not using the ISRT that @DaveCTurner wanted. If it would throw, then that's a new behavior that I can implement, but I wonder if we want that? It would mean that the promotable replicas might proceed (with the older replication group) while the unpromotables would fail because we're not on the latest cluster state. I find this discrepancy a bit weird.

kingherc committed:
Small PR comment

kingherc committed:
Broadcast request implements IndicesRequest

To appease RBACEngine

kingherc committed:
Fix checkstyle

kingherc committed:
Small serialization optimization

No need to send indices over the wire

kingherc committed:
Merge remote-tracking branch 'main' into feature/ES-5454-action-broadcast-to-unpromotables

kingherc commented:
Hi reviewers, @DaveCTurner , @tlrx , @idegtiarenko , gentle reminder for review after tackling latest comments. Thanks!

DaveCTurner reviewed (approved):
LGTM, thanks for all the iterations here @kingherc

kingherc commented:
@idegtiarenko , @tlrx if you have further comments, I can of course open a follow up PR, feel free to comment more if/as needed!
